File: daily-programmer/cpp/1-hard-guessing-game-reverse/main.cpp
#include "utils.hpp"

using namespace std;
using namespace utils;

int main()
{
    cout << "You may pick a number in [0, 100]" << newl
         << "Input [l] for lower, [h] for higher, [i] for it" << endl;

    int range_min = 0;
    int range_max = 100;

    for (int iter = 1;; iter++) {
        if (range_min == range_max) {
            cerr << "Cheater! You re-picked number in the middle of game"
                 << endl;
            exit(EXIT_FAILURE);
        }

        int guess = (range_min + range_max) / 2;

        while (true) {
            cout << "Your number is " << guess << "? ";
            string in = prompt("> ");

            if (in == "l") {
                range_max = guess;
                break;
            }

            if (in == "h") {
                range_min = guess;
                break;
            }

            if (in == "i") {
                cout << "PC won in " << iter << " iterations" << endl;
                exit(EXIT_SUCCESS);
            }

            cout << "Invalid input, please select one of [l], [h] or [i]"
                 << endl;
        }
    }
}

File: daily-programmer/1-hard-guessing-game-reverse/java/Main.java
import java.util.Scanner;

class Main {

    public static void main(String[] args) {

        new Application().run();

    }

}

class Application {

    public void run() {

        final var scanner = new Scanner(System.in);

        var rangeMin = 0;
        var rangeMax = 100;

        System.out.println("You may pick a number in [0, 100]\n"
                + "Input [l] for lower, [h] for higher, [i] for it");

        for (var iter = 1;; iter++) {

            if (rangeMin == rangeMax) {
                System.out.println("Cheater! You re-picked number in the middle of game");
                return;
            }

            final var guess = (rangeMin + rangeMax) / 2;

            while (true) {

                System.out.print("Your number is '" + guess + "'? > ");

                var input = scanner.nextLine();
                if (input.equals("l")) {
                    rangeMax = guess;
                    break;
                }

                if (input.equals("h")) {
                    rangeMin = guess;
                    break;
                }

                if (input.equals("i")) {
                    System.out.println("PC won in '" + iter + "' iterations");
                    return;
                }

                System.out.println("Invalid input, please select one of [l], [h] or [i]");

            }

        }

    }

}